### ISSUE TITLE
Exception: "This operation is only valid on generic types."

### ISSUE THREAD

**Issue by oruchreis** [salarcode/Bois#9]
Hi,
I'm trying to serialize an object that have IDictionary and IList types. But it throws at ReadDictionary method, which it doesn't check if the type is generic or not:
```
using System.Collections.Concurrent;
using System.Runtime.Serialization;
using Salar.Bois;

[DataContract]
public class Asd
{
    [DataMember]
    public string Member { get; set; }
    public string NotMember { get; set; }
    [DataMember]
    public ExchangeRates Rates { get; set; }
    [DataMember]
    public List<ExchangeRate> RateList { get; set; }
    [DataMember]
    public ConcurrentDictionary<string, ExchangeRate> RateDic { get; set; }
    [DataMember]
    public ExchangeRateCustomList RateCustomList { get; set; }
}

[DataContract]
public class ExchangeRate
{
    [DataMember]
    public double Price { get; set; }
}

[CollectionDataContract(
        Name = "ExchangeRates",
        ItemName = "ExchangeRate",
        KeyName = "Currency",
        ValueName = "Rate")]
public class ExchangeRates : ConcurrentDictionary<string, ExchangeRate>
{
    
}

[CollectionDataContract(
        Name = "ExchangeRateList",
        ItemName = "ExchangeRate",
        ValueName = "Rate")]
public class ExchangeRateCustomList: List<ExchangeRate>
{
    
}

var asd = new Asd
{
    Member = "Member",
    NotMember = "NotMember",
    Rates = new ExchangeRates
    {
        ["TRY"] = new ExchangeRate
        {
            Price = 1.4d
        }
    },
    RateList = new List<ExchangeRate>
    {
        new ExchangeRate{
            Price = 1.4444213d
        }
    },
    RateDic = new ConcurrentDictionary<string, ExchangeRate>
    {
        ["TRY"] = new ExchangeRate
        {
            Price = 1.4d
        }
    },
    RateCustomList = new ExchangeRateCustomList
    {
        new ExchangeRate{
            Price = 1.4444213d
        }
    }
};

var boisSerializer = new BoisSerializer();
using (var ms = new MemoryStream())
{
    boisSerializer.Serialize(asd, ms);
    ms.Position = 0;
    boisSerializer.Deserialize<Asd>(ms).Dump();
} 
```

**Comment by salarcode** [comment]
Version 2 doesn't support ConcurrentDictionary. Version 3 is under development and it will support it.
It should be released soon.

**Comment by salarcode** [comment]
This is fixed in 3.0 beta1
https://www.nuget.org/packages/Salar.Bois/3.0.0-beta1
